File: Java/lab17_Mediator/task_2_1 – solution/src/Mediator.java
public class Mediator {

    private ECommerceSite site;
    private Customer customer;
    private Driver driver;

    public void setSite(ECommerceSite site) {
        this.site = site;
    }

    public void setCustomer(Customer customer) {
        this.customer = customer;
    }

    public void setDriver(Driver driver) {
        this.driver = driver;
    }

    //    site = new ECommerceSite(this);
    public void buy(String item, int quantity) {
        if (site.checkInStock(item, quantity)) {
            site.sell(item, quantity);
            driver.deliver(item, quantity, customer);
        }
    }
}

File: Cpp/lab17_Mediator/task_2_1 - solution/src/mediator.cpp
#include "mediator.hpp"


void Mediator::setSite(ECommerceSite& site)
{
  this->site = &site;
}

void Mediator::setCustomer(Customer& customer)
{
  this->customer = &customer;
}

void Mediator::setDriver(Driver& driver)
{
  this->driver = &driver;
}

void Mediator::buy(std::string const& item, int quantity)
{
  if (this->site->checkInStock(item, quantity))
  {
    this->site->sell(item, quantity);
    this->driver->deliver(item, quantity, *this->customer);
  }
}
